pkg/server/internal/extensions/extensions.go

```diff
@@ -0,0 +1,167 @@
+		el = append(el, fmt.Errorf("forbidden extensions: %v", csr.Extensions))
+	}
+
+	for _, extension := range csr.Extensions {
+		switch extension.Id.String() {
+		case oidExtensionSubjectAltName.String():
+			el = append(el, validateExtensionURI(extension))
+
+		case oidExtensionKeyUsage.String():
+			el = append(el, validateKeyUsageExtension(extension.Value))
+
+		case oidExtensionExtendedKeyUsage.String():
+			el = append(el, validateExtendedKeyUsageExtension(extension))
+
+		default:
+			el = append(el, fmt.Errorf("forbidden extension: %s", extension.Id))
+		}
+	}
+
+	return utilerrors.NewAggregate(el)
+}
+
+// validateKeyUsageExtension validates that the passed extension value contains
+// accepted key usages
+func validateKeyUsageExtension(value []byte) error {
+	if len(value) != 4 {
+		return fmt.Errorf("forbidden key usage: %v", value)
+	}
+
+	extValue := make([]byte, len(value))
+	copy(extValue, value)
+	for _, usage := range allowedKeyUsages {
+		for i, b := range usage {
+			extValue[i] = extValue[i] &^ b
+		}
+	}
+
+	if !bytes.Equal(extValue, []byte{0, 0, 0, 0}) {
+		return fmt.Errorf("forbidden key usage: %v", value)
+	}
+
+	return nil
+}
+
+// validateExtendedKeyUsageExtension validates that the passed extension
+// contains accepted extended key usages
+func validateExtendedKeyUsageExtension(extension pkix.Extension) error {
+	if !extension.Id.Equal(oidExtensionExtendedKeyUsage) {
+		return fmt.Errorf("non extended key usage extension: %s", extension.Id)
+	}
+
+	var asn1ExtendedUsages []asn1.ObjectIdentifier
+	_, err := asn1.Unmarshal(extension.Value, &asn1ExtendedUsages)
+	if err != nil {
+		return fmt.Errorf("failed to parse extended key usages: %s", err)
+	}
+
+	var el []error
+	for _, usage := range asn1ExtendedUsages {
+		if !usage.Equal(oidExtKeyUsageClientAuth) && !usage.Equal(oidExtKeyUsageServerAuth) {
+			el = append(el, fmt.Errorf("forbidden extended key usage: %s", usage))
+		}
+	}
+
+	return utilerrors.NewAggregate(el)
+}
+
+// validateExtensionURI validates that the passed extension is a correctly
+// encoded URI SAN
+func validateExtensionURI(ext pkix.Extension) error {
+	if !ext.Id.Equal(oidExtensionSubjectAltName) {
+		return fmt.Errorf("extension is not a SAN type: %s", ext.Id)
+	}
+	var sequence asn1.RawValue
+	if rest, err := asn1.Unmarshal(ext.Value, &sequence); err != nil {
+		return fmt.Errorf("failed to unmarshal san extension: %v", err)
+	} else if len(rest) != 0 {
+		return fmt.Errorf("san extension incorrectly encoded: %v", ext.Value)
+	}
+
+	// Check the rawValue is a sequence.
+	if !sequence.IsCompound || sequence.Tag != asn1.TagSequence || sequence.Class != asn1.ClassUniversal {
+		return fmt.Errorf("san extension is incorrectly encoded: %v", ext.Value)
+	}
+
+	for bytes := sequence.Bytes; len(bytes) > 0; {
+		var (
+			rawValue asn1.RawValue
+			err      error
+		)
+
+		bytes, err = asn1.Unmarshal(bytes, &rawValue)
+		if err != nil {
+			return err
+		}
+
+		if rawValue.Tag != asn1TagURI {
```

Comment:
For TLS I'd normally expect at least `dNSName` (identifier 2) to be permitted here. Does istio require _only_ URIs? If so, it might be good to add a comment to that effect

Reply:
Correct, istio is URI _only_ as far as I'm aware and can think of. The only cert which includes DNS is the istiod certificate (which it needs for its k8s webhook) which is minted through a `Certificate` resource rather than istio-csr. If and when istiod adds support for using the same mechanism to get its certificate, we will need to add a specific allow here for only that identity.

Reply:
good to know, thanks :+1: